File: apps/syn-cli/src/syn_cli/commands/watch.py
"""Watch commands — live SSE stream of execution or global activity."""

from __future__ import annotations

import json
from typing import Annotated

import typer

from syn_cli._output import console, format_cost, format_timestamp, print_error, status_style
from syn_cli.client import get_api_url, get_streaming_client

app = typer.Typer(
    name="watch",
    help="Watch live execution or activity streams",
    no_args_is_help=True,
)


def _parse_sse_line(line: str) -> dict[str, object] | None:
    """Parse a single SSE data line into a dict, or None if not a data line."""
    if line.startswith("data: "):
        raw = line[6:].strip()
        if raw:
            try:
                parsed: dict[str, object] = json.loads(raw)
                return parsed
            except json.JSONDecodeError:
                pass
    return None


def _render_event(frame: dict[str, object]) -> None:
    """Render an SSE event frame to the console."""
    event_type = str(frame.get("event_type", ""))
    data = frame.get("data")
    if not isinstance(data, dict):
        data = {}

    ts = format_timestamp(str(frame.get("timestamp", "")))

    if event_type == "ExecutionStarted":
        exec_id = str(data.get("execution_id", ""))[:12]
        console.print(f"[dim]{ts}[/dim]  [green]▶ started[/green]  {exec_id}")
    elif event_type in ("ExecutionCompleted", "WorkflowCompleted"):
        exec_id = str(data.get("execution_id", ""))[:12]
        cost = format_cost(data.get("total_cost_usd") or "0")
        console.print(f"[dim]{ts}[/dim]  [green]✓ completed[/green]  {exec_id}  {cost}")
    elif event_type in ("ExecutionFailed", "WorkflowFailed"):
        exec_id = str(data.get("execution_id", ""))[:12]
        err = str(data.get("error_message") or "")[:60]
        console.print(f"[dim]{ts}[/dim]  [red]✗ failed[/red]  {exec_id}  {err}")
    elif event_type == "PhaseStarted":
        phase = str(data.get("phase_name", ""))
        console.print(f"[dim]{ts}[/dim]  [blue]→ phase[/blue]  {phase}")
    elif event_type == "PhaseCompleted":
        phase = str(data.get("phase_name", ""))
        console.print(f"[dim]{ts}[/dim]  [blue]✓ phase[/blue]  {phase}")
    elif event_type == "SessionTokensRecorded":
        tokens = data.get("total_tokens", 0)
        cost = format_cost(data.get("cost_usd") or "0")
        console.print(f"[dim]{ts}[/dim]  [dim]tokens {tokens}  {cost}[/dim]")
    elif event_type:
        status = str(data.get("status", ""))
        style = status_style(status)
        suffix = f"  [{style}]{status}[/{style}]" if status and style else ""
        console.print(f"[dim]{ts}[/dim]  [dim]{event_type}[/dim]{suffix}")


@app.command("execution")
def watch_execution(
    execution_id: Annotated[str, typer.Argument(help="Execution ID to watch")],
) -> None:
    """Stream live events for a specific execution."""
    url = f"/sse/executions/{execution_id}"
    base = get_api_url()
    console.print(f"[dim]Connecting to {base}{url} …[/dim]")
    console.print("[dim]Press Ctrl-C to stop.[/dim]")
    console.print()

    try:
        with get_streaming_client() as client:
            with client.stream("GET", url) as response:
                if response.status_code == 404:
                    print_error(f"Execution not found: {execution_id}")
                    raise typer.Exit(1)
                if response.status_code != 200:
                    print_error(f"HTTP {response.status_code}")
                    raise typer.Exit(1)

                for line in response.iter_lines():
                    if not line:
                        continue
                    frame = _parse_sse_line(line)
                    if frame is None:
                        continue
                    frame_type = str(frame.get("type", ""))
                    if frame_type == "connected":
                        console.print("[green]Connected.[/green]")
                    elif frame_type == "terminal":
                        _render_event(frame)
                        console.print("[dim]Stream ended.[/dim]")
                        return
                    elif frame_type == "event":
                        _render_event(frame)
    except KeyboardInterrupt:
        console.print("\n[dim]Stopped.[/dim]")
    except Exception:
        print_error(f"Could not connect to API at {get_api_url()}")
        console.print("[dim]Make sure the API server is running.[/dim]")
        raise typer.Exit(1)


@app.command("activity")
def watch_activity() -> None:
    """Stream live global activity across all executions."""
    url = "/sse/activity"
    base = get_api_url()
    console.print(f"[dim]Connecting to {base}{url} …[/dim]")
    console.print("[dim]Press Ctrl-C to stop.[/dim]")
    console.print()

    try:
        with get_streaming_client() as client:
            with client.stream("GET", url) as response:
                if response.status_code != 200:
                    print_error(f"HTTP {response.status_code}")
                    raise typer.Exit(1)

                for line in response.iter_lines():
                    if not line:
                        continue
                    frame = _parse_sse_line(line)
                    if frame is None:
                        continue
                    frame_type = str(frame.get("type", ""))
                    if frame_type == "connected":
                        console.print("[green]Connected — watching global activity.[/green]")
                    elif frame_type == "event":
                        _render_event(frame)
    except KeyboardInterrupt:
        console.print("\n[dim]Stopped.[/dim]")
    except Exception:
        print_error(f"Could not connect to API at {get_api_url()}")
        console.print("[dim]Make sure the API server is running.[/dim]")
        raise typer.Exit(1)

File: apps/syn-cli-node/src/commands/watch.ts
/**
 * Watch commands — live SSE streaming for executions and global activity.
 * Port of apps/syn-cli/src/syn_cli/commands/watch/_sse.py + _render.py
 */

import { CommandGroup, type CommandDef, type ParsedArgs } from "../framework/command.js";
import { CLIError } from "../framework/errors.js";
import { streamSSE, type SSEEvent } from "../client/sse.js";
import { print, printError, printDim } from "../output/console.js";
import { style, BOLD, CYAN, DIM, GREEN, RED, YELLOW } from "../output/ansi.js";
import { formatCost, formatTimestamp, formatTokens } from "../output/format.js";

const EVENT_STYLES: Record<string, string> = {
  WorkflowExecutionStarted: GREEN,
  WorkflowCompleted: GREEN,
  WorkflowFailed: RED,
  PhaseStarted: CYAN,
  PhaseCompleted: GREEN,
  PhaseFailed: RED,
  SessionTokensRecorded: DIM,
  ExecutionPaused: YELLOW,
  ExecutionResumed: GREEN,
  ExecutionCancelled: RED,
};

function extractEventId(data: Record<string, unknown>): string {
  return (data["workflow_execution_id"] ?? data["execution_id"] ?? data["session_id"] ?? "") as string;
}

function renderEventData(parts: string[], data: Record<string, unknown>): void {
  const id = extractEventId(data);
  if (id) parts.push(` ${style(id.slice(0, 12), DIM)}`);
  if (data["workflow_name"]) parts.push(` ${String(data["workflow_name"])}`);
  if (data["phase_name"]) parts.push(` phase:${String(data["phase_name"])}`);
  if (data["total_tokens"]) parts.push(` tokens:${formatTokens(Number(data["total_tokens"]))}`);
  if (data["cost_usd"]) parts.push(` cost:${formatCost(String(data["cost_usd"]))}`);
  if (data["error_message"]) parts.push(` ${style(String(data["error_message"]).slice(0, 80), RED)}`);
}

function renderEvent(event: SSEEvent): void {
  const ts = event.timestamp ? formatTimestamp(event.timestamp) : style("now", DIM);
  const eventType = event.event_type ?? event.type ?? "unknown";
  const eventStyle = EVENT_STYLES[eventType] ?? "";

  const parts = [style(ts, DIM), " ", eventStyle ? style(eventType, eventStyle) : eventType];

  if (event.data) {
    renderEventData(parts, event.data);
  }

  print(parts.join(""));
}

const executionCommand: CommandDef = {
  name: "execution",
  description: "Stream live events for a specific execution",
  args: [{ name: "execution-id", description: "Execution ID to watch", required: true }],
  handler: async (parsed: ParsedArgs) => {
    const id = parsed.positionals[0];
    if (!id) { printError("Missing execution-id"); throw new CLIError("Missing argument", 1); }

    print(style(`Watching execution ${id}...`, BOLD));
    printDim("Press Ctrl+C to stop.\n");

    try {
      for await (const event of streamSSE(`/sse/executions/${id}`)) {
        renderEvent(event);
      }
      print(style("\nStream ended.", DIM));
    } catch (err) {
      if (err instanceof CLIError) throw err;
      if (err instanceof Error && err.name === "AbortError") return;
      printError(`Stream error: ${err instanceof Error ? err.message : String(err)}`);
      throw new CLIError("Stream failed", 1);
    }
  },
};

const activityCommand: CommandDef = {
  name: "activity",
  description: "Stream live global activity across all executions",
  handler: async () => {
    print(style("Watching global activity...", BOLD));
    printDim("Press Ctrl+C to stop.\n");

    try {
      for await (const event of streamSSE("/watch/activity")) {
        renderEvent(event);
      }
      print(style("\nStream ended.", DIM));
    } catch (err) {
      if (err instanceof CLIError) throw err;
      if (err instanceof Error && err.name === "AbortError") return;
      printError(`Stream error: ${err instanceof Error ? err.message : String(err)}`);
      throw new CLIError("Stream failed", 1);
    }
  },
};

export const watchGroup = new CommandGroup("watch", "Stream live execution events via SSE");
watchGroup.command(executionCommand).command(activityCommand);
